src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/ManagedIndexRunner.kt

```diff
@@ -15,20 +15,92 @@
 
 package com.amazon.opendistroforelasticsearch.indexstatemanagement
 
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.action.updateindexmetadata.UpdateManagedIndexMetaDataAction
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.action.updateindexmetadata.UpdateManagedIndexMetaDataRequest
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.elasticapi.getIndexMetadata
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.models.ManagedIndexConfig
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.models.ManagedIndexMetaData
 import com.amazon.opendistroforelasticsearch.jobscheduler.spi.JobExecutionContext
 import com.amazon.opendistroforelasticsearch.jobscheduler.spi.ScheduledJobParameter
 import com.amazon.opendistroforelasticsearch.jobscheduler.spi.ScheduledJobRunner
 import org.apache.logging.log4j.LogManager
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.index.Index
+import org.elasticsearch.threadpool.ThreadPool
 
-class ManagedIndexRunner : ScheduledJobRunner {
+object ManagedIndexRunner : ScheduledJobRunner {
 
     private val logger = LogManager.getLogger(javaClass)
 
-    companion object {
-        val instance = ManagedIndexRunner()
+    private lateinit var clusterService: ClusterService
+    private lateinit var client: Client
+    private lateinit var threadPool: ThreadPool
+
+    fun registerClusterService(clusterService: ClusterService): ManagedIndexRunner {
+        this.clusterService = clusterService
+        return this
+    }
+
+    fun registerClient(client: Client): ManagedIndexRunner {
+        this.client = client
+        return this
+    }
+
+    // TODO we may not need ThreadPool if we use Kotlin Coroutine. This is just temporary implementation for simplicity.
+    fun registerThreadPool(threadPool: ThreadPool): ManagedIndexRunner {
+        this.threadPool = threadPool
+        return this
     }
 
     override fun runJob(job: ScheduledJobParameter, context: JobExecutionContext) {
         logger.info("runJob for ${job.name}")
+
+        // TODO Use Kotlin Coroutine. This just temporary implementation.
+        val runnable = Runnable {
+            managedIndexRun(job, context)
+        }
+
+        threadPool.generic().submit(runnable)
+    }
+
+    private fun managedIndexRun(job: ScheduledJobParameter, context: JobExecutionContext) {
+        context.expectedExecutionTime
+        if (job !is ManagedIndexConfig) {
+            logger.info("Cannot run job for none ManagedIndexConfig")
```

Comment:
Updated to remove casting. https://kotlinlang.org/docs/reference/typecasts.html this document shows that if we return/throw error in the if-statement the code below the if-statement will do a smart cast and know the job type is `ManagedIndexConfig`.

Reply:
This check should actually be done in the runJob and the runManagedIndexJob should accept a job: ManagedIndexConfig parameter. Was going to update this later with coroutine switch since this file has a lot of TODOs